### ISSUE TITLE
Johnsons shortest path algorithm bug

### ISSUE THREAD

**Issue by jkinable** [jgrapht/jgrapht#408]
There seems to be a bug in Johnsons shortest path algorithm.
The following test case causes a Vertex Factory Failed exception. @d-michail could you have a look at this?

```
@Test
public void testDisconnectedGraph(){
    Graph<Integer, DefaultEdge> graph = new SimpleDirectedGraph<>(DefaultEdge.class);
    Graphs.addAllVertices(graph, Arrays.asList(0,1,2,3,4,5,6));
    graph.addEdge(0,1);
    graph.addEdge(1,2);
    graph.addEdge(2,3);
    graph.addEdge(3,0);

    graph.addEdge(4,5);
    graph.addEdge(5,6);
    graph.addEdge(6,4);

    JohnsonShortestPaths<Integer, DefaultEdge> sp =
	    new JohnsonShortestPaths<>(graph, Integer.class);

    assertEquals(2.0, sp.getPathWeight(0,2), 0.0);
    assertEquals(1.0, sp.getPathWeight(4,5), 0.0);
    assertTrue(Double.isInfinite(sp.getPathWeight(3, 4)));
}
```

**Comment by d-michail** [comment]
This is not a bug. 

Johnson's algorithm needs to be able to create new vertices which are not already in the graph. The provided `VertexFactory` which is derived from the `Integer.class` returns vertices which are already in the graph such as 0 to 6.

You need to provide a `VertexFactory` which creates vertices which are larger than 6.

**Comment by jkinable** [comment]
hm good point. This isn't particularly intuitive from the documentation though.

Aside from that, changing test case above:
```
JohnsonShortestPaths<Integer, DefaultEdge> sp =
                new JohnsonShortestPaths<>(graph, new IntegerVertexFactory(7));
```
produces a Nullpointer exception.

**Comment by d-michail** [comment]
That looks like a bug. I will take a look.

**Comment by jkinable** [comment]
Fixed in #409
